File: sdk/provider/spacewave/spacewave.ts
import { Provider } from '../provider.js'
import { ClientResourceRef } from '@aptre/bldr-sdk/resource/client.js'
import {
  SpacewaveProviderResourceService,
  SpacewaveProviderResourceServiceClient,
} from './spacewave_srpc.pb.js'
import type {
  CloudProviderConfig,
  ConfirmDesktopPasskeyRequest,
  ConfirmDesktopPasskeyResponse,
  ConfirmDesktopSSORequest,
  ConfirmDesktopSSOResponse,
  CreateAccountRequest,
  CreateAccountResponse,
  GenerateAuthKeypairsResponse,
  GeneratePasskeyPrfSaltResponse,
  GetCloudProviderConfigRequest,
  GetLinkedCloudSessionResponse,
  LoginAccountRequest,
  LoginAccountResponse,
  LoginOrCreateAccountRequest,
  LoginOrCreateAccountResponse,
  LoginWithEntityKeyResponse,
  PasskeyCheckUsernameRequest,
  PasskeyCheckUsernameResponse,
  PasskeyConfirmSignupRequest,
  PasskeyConfirmSignupResponse,
  PasskeyAuthOptionsResponse,
  PasskeyAuthOptionsRequest,
  PasskeyAuthVerifyRequest,
  PasskeyAuthVerifyResponse,
  PasskeyRegisterChallengeRequest,
  PasskeyRegisterChallengeResponse,
  RequestRecoveryEmailRequest,
  RequestRecoveryEmailResponse,
  ReauthenticateSessionRequest,
  ReauthenticateSessionResponse,
  RecoverExecuteRequest,
  SSONonceExchangeRequest,
  SSOCodeExchangeRequest,
  SSOCodeExchangeResponse,
  RecoverExecuteResponse,
  RecoverVerifyResponse,
  StartBrowserHandoffRequest,
  StartBrowserHandoffResponse,
  StartDesktopPasskeyRequest,
  StartDesktopPasskeyResponse,
  StartDesktopSSORequest,
  StartDesktopSSOResponse,
  UnwrapWithPasskeyPrfRequest,
  UnwrapWithPasskeyPrfResponse,
  UnwrapPemWithPinResponse,
  WrapWithPasskeyPrfRequest,
  WrapWithPasskeyPrfResponse,
  WrapPemWithPinResponse,
} from './spacewave.pb.js'

// SpacewaveProvider wraps a Provider resource with spacewave-specific pre-auth RPCs.
// Post-auth session-scoped RPCs have moved to SpacewaveSession
// (sdk/session/spacewave-session.ts), accessed via session.spacewave.
export class SpacewaveProvider extends Provider {
  private swService: SpacewaveProviderResourceService

  constructor(resourceRef: ClientResourceRef) {
    super(resourceRef)
    this.swService = new SpacewaveProviderResourceServiceClient(
      resourceRef.client,
    )
  }

  // createAccount creates an account on the spacewave provider.
  public async createAccount(
    request: CreateAccountRequest,
    abortSignal?: AbortSignal,
  ): Promise<CreateAccountResponse> {
    return await this.swService.CreateAccount(request, abortSignal)
  }

  // loginAccount attempts to log in without creating an account.
  public async loginAccount(
    request: LoginAccountRequest,
    abortSignal?: AbortSignal,
  ): Promise<LoginAccountResponse> {
    return await this.swService.LoginAccount(request, abortSignal)
  }

  // loginOrCreateAccount logs in or creates an account on the spacewave provider.
  public async loginOrCreateAccount(
    request: LoginOrCreateAccountRequest,
    abortSignal?: AbortSignal,
  ): Promise<LoginOrCreateAccountResponse> {
    return await this.swService.LoginOrCreateAccount(request, abortSignal)
  }

  // loginWithEntityKey creates a session using a pre-resolved entity PEM key.
  public async loginWithEntityKey(
    pemPrivateKey: Uint8Array,
    abortSignal?: AbortSignal,
  ): Promise<LoginWithEntityKeyResponse> {
    return await this.swService.LoginWithEntityKey(
      { pemPrivateKey },
      abortSignal,
    )
  }

  // generateAuthKeypairs creates account and session identity key material.
  public async generateAuthKeypairs(
    abortSignal?: AbortSignal,
  ): Promise<GenerateAuthKeypairsResponse> {
    return await this.swService.GenerateAuthKeypairs({}, abortSignal)
  }

  // wrapPemWithPin wraps an entity PEM with PIN encryption.
  public async wrapPemWithPin(
    pemPrivateKey: string,
    pin: string,
    abortSignal?: AbortSignal,
  ): Promise<WrapPemWithPinResponse> {
    return await this.swService.WrapPemWithPin(
      { pemPrivateKey, pin },
      abortSignal,
    )
  }

  // unwrapPemWithPin unwraps an entity PEM with PIN encryption.
  public async unwrapPemWithPin(
    wrappedPemBase64: string,
    pin: string,
    abortSignal?: AbortSignal,
  ): Promise<UnwrapPemWithPinResponse> {
    return await this.swService.UnwrapPemWithPin(
      { wrappedPemBase64, pin },
      abortSignal,
    )
  }

  // generatePasskeyPrfSalt creates a provider-owned WebAuthn PRF salt.
  public async generatePasskeyPrfSalt(
    abortSignal?: AbortSignal,
  ): Promise<GeneratePasskeyPrfSaltResponse> {
    return await this.swService.GeneratePasskeyPrfSalt({}, abortSignal)
  }

  // wrapWithPasskeyPrf wraps an auth blob with passkey PRF encryption.
  public async wrapWithPasskeyPrf(
    request: WrapWithPasskeyPrfRequest,
    abortSignal?: AbortSignal,
  ): Promise<WrapWithPasskeyPrfResponse> {
    return await this.swService.WrapWithPasskeyPrf(request, abortSignal)
  }

  // unwrapWithPasskeyPrf unwraps an auth blob with passkey PRF encryption.
  public async unwrapWithPasskeyPrf(
    request: UnwrapWithPasskeyPrfRequest,
    abortSignal?: AbortSignal,
  ): Promise<UnwrapWithPasskeyPrfResponse> {
    return await this.swService.UnwrapWithPasskeyPrf(request, abortSignal)
  }

  // passkeyCheckUsername acknowledges the opaque first passkey step.
  public async passkeyCheckUsername(
    request: PasskeyCheckUsernameRequest,
    abortSignal?: AbortSignal,
  ): Promise<PasskeyCheckUsernameResponse> {
    return await this.swService.PasskeyCheckUsername(request, abortSignal)
  }

  // passkeyRegisterChallenge fetches WebAuthn registration options for signup.
  public async passkeyRegisterChallenge(
    request: PasskeyRegisterChallengeRequest,
    abortSignal?: AbortSignal,
  ): Promise<PasskeyRegisterChallengeResponse> {
    return await this.swService.PasskeyRegisterChallenge(request, abortSignal)
  }

  // passkeyAuthOptions fetches WebAuthn authentication options from the cloud.
  public async passkeyAuthOptions(
    request: PasskeyAuthOptionsRequest = {},
    abortSignal?: AbortSignal,
  ): Promise<PasskeyAuthOptionsResponse> {
    return await this.swService.PasskeyAuthOptions(request, abortSignal)
  }

  // passkeyAuthVerify verifies a WebAuthn authentication credential.
  public async passkeyAuthVerify(
    request: PasskeyAuthVerifyRequest,
    abortSignal?: AbortSignal,
  ): Promise<PasskeyAuthVerifyResponse> {
    return await this.swService.PasskeyAuthVerify(request, abortSignal)
  }

  // passkeyConfirmSignup confirms browser-owned passkey signup for the web flow.
  public async passkeyConfirmSignup(
    request: PasskeyConfirmSignupRequest,
    abortSignal?: AbortSignal,
  ): Promise<PasskeyConfirmSignupResponse> {
    return await this.swService.PasskeyConfirmSignup(request, abortSignal)
  }

  // ssoCodeExchange exchanges an OAuth authorization code for account info.
  public async ssoCodeExchange(
    request: SSOCodeExchangeRequest,
    abortSignal?: AbortSignal,
  ): Promise<SSOCodeExchangeResponse> {
    return await this.swService.SSOCodeExchange(request, abortSignal)
  }

  // ssoNonceExchange exchanges an auth-session nonce for stored SSO result.
  public async ssoNonceExchange(
    request: SSONonceExchangeRequest,
    abortSignal?: AbortSignal,
  ): Promise<SSOCodeExchangeResponse> {
    return await this.swService.SSONonceExchange(request, abortSignal)
  }

  // getCloudProviderConfig fetches pre-auth provider config.
  public async getCloudProviderConfig(
    abortSignal?: AbortSignal,
  ): Promise<CloudProviderConfig> {
    return await this.swService.GetCloudProviderConfig(
      {} satisfies GetCloudProviderConfigRequest,
      abortSignal,
    )
  }

  // startBrowserHandoff opens the browser auth handoff flow on native clients.
  public async startBrowserHandoff(
    request: StartBrowserHandoffRequest,
    abortSignal?: AbortSignal,
  ): Promise<StartBrowserHandoffResponse> {
    return await this.swService.StartBrowserHandoff(request, abortSignal)
  }

  // startDesktopSSO starts the native desktop SSO flow.
  public async startDesktopSSO(
    request: StartDesktopSSORequest,
    abortSignal?: AbortSignal,
  ): Promise<StartDesktopSSOResponse> {
    return await this.swService.StartDesktopSSO(request, abortSignal)
  }

  // confirmDesktopSSO completes native desktop SSO account creation.
  public async confirmDesktopSSO(
    request: ConfirmDesktopSSORequest,
    abortSignal?: AbortSignal,
  ): Promise<ConfirmDesktopSSOResponse> {
    return await this.swService.ConfirmDesktopSSO(request, abortSignal)
  }

  // confirmSSO completes browser or native SSO account creation.
  public async confirmSSO(
    request: ConfirmDesktopSSORequest,
    abortSignal?: AbortSignal,
  ): Promise<ConfirmDesktopSSOResponse> {
    return await this.swService.ConfirmDesktopSSO(request, abortSignal)
  }

  // startDesktopPasskey starts the native desktop passkey flow.
  public async startDesktopPasskey(
    request: StartDesktopPasskeyRequest,
    abortSignal?: AbortSignal,
  ): Promise<StartDesktopPasskeyResponse> {
    return await this.swService.StartDesktopPasskey(request, abortSignal)
  }

  // confirmDesktopPasskey completes native desktop passkey account creation.
  public async confirmDesktopPasskey(
    request: ConfirmDesktopPasskeyRequest,
    abortSignal?: AbortSignal,
  ): Promise<ConfirmDesktopPasskeyResponse> {
    return await this.swService.ConfirmDesktopPasskey(request, abortSignal)
  }

  // requestRecoveryEmail requests a recovery email for an account.
  public async requestRecoveryEmail(
    request: RequestRecoveryEmailRequest,
    abortSignal?: AbortSignal,
  ): Promise<RequestRecoveryEmailResponse> {
    return await this.swService.RequestRecoveryEmail(request, abortSignal)
  }

  // recoverVerify verifies a recovery token from an email link.
  public async recoverVerify(
    token: string,
    abortSignal?: AbortSignal,
  ): Promise<RecoverVerifyResponse> {
    return await this.swService.RecoverVerify({ token }, abortSignal)
  }

  // recoverExecute completes account recovery by deriving a new password keypair.
  public async recoverExecute(
    request: RecoverExecuteRequest,
    abortSignal?: AbortSignal,
  ): Promise<RecoverExecuteResponse> {
    return await this.swService.RecoverExecute(request, abortSignal)
  }

  // getLinkedCloudSession returns the linked cloud session index if it exists.
  public async getLinkedCloudSession(
    abortSignal?: AbortSignal,
  ): Promise<GetLinkedCloudSessionResponse> {
    return await this.swService.GetLinkedCloudSession({}, abortSignal)
  }

  // reauthenticateSession re-authenticates a session whose key became stale.
  public async reauthenticateSession(
    request: ReauthenticateSessionRequest,
    abortSignal?: AbortSignal,
  ): Promise<ReauthenticateSessionResponse> {
    return await this.swService.ReauthenticateSession(request, abortSignal)
  }
}

File: sdk/provider/spacewave/spacewave.go
package s4wave_provider_spacewave

import (
	"context"

	resource_client "github.com/s4wave/spacewave/bldr/resource/client"
)

// SpacewaveProvider is the SDK wrapper for the SpacewaveProviderResourceService.
type SpacewaveProvider struct {
	client  *resource_client.Client
	ref     resource_client.ResourceRef
	service SRPCSpacewaveProviderResourceServiceClient
}

// NewSpacewaveProvider creates a new SpacewaveProvider resource wrapper.
func NewSpacewaveProvider(client *resource_client.Client, ref resource_client.ResourceRef) (*SpacewaveProvider, error) {
	srpcClient, err := ref.GetClient()
	if err != nil {
		return nil, err
	}
	return &SpacewaveProvider{
		client:  client,
		ref:     ref,
		service: NewSRPCSpacewaveProviderResourceServiceClient(srpcClient),
	}, nil
}

// GetResourceRef returns the resource reference.
func (s *SpacewaveProvider) GetResourceRef() resource_client.ResourceRef {
	return s.ref
}

// Release releases the resource reference.
func (s *SpacewaveProvider) Release() {
	s.ref.Release()
}

// CreateAccount creates an account on the spacewave provider.
func (s *SpacewaveProvider) CreateAccount(ctx context.Context, entityID, password, turnstileToken string) (*CreateAccountResponse, error) {
	return s.service.CreateAccount(ctx, &CreateAccountRequest{
		EntityId:       entityID,
		TurnstileToken: turnstileToken,
		Credential: &CreateAccountRequest_Password{
			Password: &PasswordCredential{Password: password},
		},
	})
}

// LoginOrCreateAccount logs in or creates an account on the spacewave provider.
func (s *SpacewaveProvider) LoginOrCreateAccount(ctx context.Context, username, password string) (*LoginOrCreateAccountResponse, error) {
	return s.service.LoginOrCreateAccount(ctx, &LoginOrCreateAccountRequest{
		Username: username,
		Password: password,
	})
}

// LoginWithEntityKey creates a session using a pre-resolved entity private key.
func (s *SpacewaveProvider) LoginWithEntityKey(ctx context.Context, pemPrivateKey []byte) (*LoginWithEntityKeyResponse, error) {
	return s.service.LoginWithEntityKey(ctx, &LoginWithEntityKeyRequest{
		PemPrivateKey: pemPrivateKey,
	})
}

// PasskeyCheckUsername acknowledges the opaque first passkey step.
func (s *SpacewaveProvider) PasskeyCheckUsername(ctx context.Context, username string) (*PasskeyCheckUsernameResponse, error) {
	return s.service.PasskeyCheckUsername(ctx, &PasskeyCheckUsernameRequest{
		Username: username,
	})
}

// PasskeyRegisterChallenge fetches WebAuthn registration options for signup.
func (s *SpacewaveProvider) PasskeyRegisterChallenge(ctx context.Context, username string) (*PasskeyRegisterChallengeResponse, error) {
	return s.service.PasskeyRegisterChallenge(ctx, &PasskeyRegisterChallengeRequest{
		Username: username,
	})
}

// PasskeyAuthOptions fetches WebAuthn authentication options from the cloud.
func (s *SpacewaveProvider) PasskeyAuthOptions(ctx context.Context, username string) (*PasskeyAuthOptionsResponse, error) {
	return s.service.PasskeyAuthOptions(ctx, &PasskeyAuthOptionsRequest{
		Username: username,
	})
}

// PasskeyAuthVerify verifies a WebAuthn authentication credential with the cloud.
func (s *SpacewaveProvider) PasskeyAuthVerify(ctx context.Context, credentialJSON string) (*PasskeyAuthVerifyResponse, error) {
	return s.service.PasskeyAuthVerify(ctx, &PasskeyAuthVerifyRequest{
		CredentialJson: credentialJSON,
	})
}

// PasskeyConfirmSignup confirms browser-owned passkey signup for the web flow.
func (s *SpacewaveProvider) PasskeyConfirmSignup(ctx context.Context, req *PasskeyConfirmSignupRequest) (*PasskeyConfirmSignupResponse, error) {
	return s.service.PasskeyConfirmSignup(ctx, req)
}

// RelayDesktopPasskey relays a browser ceremony result back to native alpha.
func (s *SpacewaveProvider) RelayDesktopPasskey(ctx context.Context, req *RelayDesktopPasskeyRequest) (*RelayDesktopPasskeyResponse, error) {
	return s.service.RelayDesktopPasskey(ctx, req)
}

// RequestRecoveryEmail requests a recovery email for an account.
func (s *SpacewaveProvider) RequestRecoveryEmail(ctx context.Context, email, turnstileToken string) (*RequestRecoveryEmailResponse, error) {
	return s.service.RequestRecoveryEmail(ctx, &RequestRecoveryEmailRequest{
		Email:          email,
		TurnstileToken: turnstileToken,
	})
}

// RecoverVerify verifies a recovery token from an email link.
func (s *SpacewaveProvider) RecoverVerify(ctx context.Context, token string) (*RecoverVerifyResponse, error) {
	return s.service.RecoverVerify(ctx, &RecoverVerifyRequest{
		Token: token,
	})
}

// RecoverExecute completes account recovery.
func (s *SpacewaveProvider) RecoverExecute(ctx context.Context, req *RecoverExecuteRequest) (*RecoverExecuteResponse, error) {
	return s.service.RecoverExecute(ctx, req)
}

// GetLinkedCloudSession returns the session index of the linked cloud session.
func (s *SpacewaveProvider) GetLinkedCloudSession(ctx context.Context) (*GetLinkedCloudSessionResponse, error) {
	return s.service.GetLinkedCloudSession(ctx, &GetLinkedCloudSessionRequest{})
}

// StartBrowserHandoff opens the browser auth handoff flow on native clients.
func (s *SpacewaveProvider) StartBrowserHandoff(ctx context.Context, req *StartBrowserHandoffRequest) (*StartBrowserHandoffResponse, error) {
	return s.service.StartBrowserHandoff(ctx, req)
}

// StartDesktopSSO starts the native desktop SSO flow.
func (s *SpacewaveProvider) StartDesktopSSO(ctx context.Context, ssoProvider string) (*StartDesktopSSOResponse, error) {
	return s.service.StartDesktopSSO(ctx, &StartDesktopSSORequest{
		SsoProvider: ssoProvider,
	})
}

// ConfirmDesktopSSO completes native desktop SSO account creation.
func (s *SpacewaveProvider) ConfirmDesktopSSO(ctx context.Context, req *ConfirmDesktopSSORequest) (*ConfirmDesktopSSOResponse, error) {
	return s.service.ConfirmDesktopSSO(ctx, req)
}

// StartDesktopPasskey starts the native desktop passkey flow.
func (s *SpacewaveProvider) StartDesktopPasskey(ctx context.Context) (*StartDesktopPasskeyResponse, error) {
	return s.service.StartDesktopPasskey(ctx, &StartDesktopPasskeyRequest{})
}

// ConfirmDesktopPasskey completes native desktop passkey account creation.
func (s *SpacewaveProvider) ConfirmDesktopPasskey(ctx context.Context, req *ConfirmDesktopPasskeyRequest) (*ConfirmDesktopPasskeyResponse, error) {
	return s.service.ConfirmDesktopPasskey(ctx, req)
}
